src/preproc/preprocessing_pass_registry.cpp

```diff
@@ -0,0 +1,56 @@
+/*********************                                                        */
+/*! \file preprocessing_pass_registry.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **  Justin Xu
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Preprocessing pass registry
+ **
+ ** Preprocessing pass registry that uses a name to store instances of
+ ** classes in a map. Makes sure passes are properly disposed of when
+ ** registry is destroyed.  
+ **/
+#include "preproc/preprocessing_pass_registry.h"
+#include <utility>
+#include <cassert>
+#include "base/output.h"
+#include "preproc/preprocessing_pass.h"
+
+namespace CVC4 {
+namespace preproc {
+
+void PreprocessingPassRegistry::init(PreprocessingPassAPI* api) {
+  for (std::pair<std::string, PreprocessingPass*>&& element :
+       d_stringToPreprocessingPass) {
+    element.second->init(api);
+  }
+}
+
+void PreprocessingPassRegistry::registerPass(
+    const std::string& ppName, PreprocessingPass* preprocessingPass) {
+  Debug("pp-registry") << "Registering pass " << ppName << std::endl;
+  Assert(d_stringToPreprocessingPass[ppName] == NULL);
```

Comment:
I'd change the check to `d_stringToPreprocessingPass.find(ppName) == d_stringToPreprocessingPass.end()`, because that communicates a bit more clearly what you actually want to check for (`d_stringToPreprocessingPass[ppName] == NULL` could be true if there was a pass `ppName` that was set to `NULL`, which shouldn't happen in our case but the reader doesn't necessarily know that). Seeing your assert actually made me think that we could also have an additional `Assert(preprocessingPass != NULL);`.